core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go

```diff
@@ -499,18 +506,53 @@ func logToTrigger(log logpoller.Log) ocr2keepers.Trigger {
 }
 
 func (r *logRecoverer) clean(ctx context.Context) {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-
-	cleaned := 0
+	r.lock.RLock()
+	var ids []string
 	for id, t := range r.visited {
-		if time.Since(t) > RecoveryCacheTTL {
-			delete(r.visited, id)
-			cleaned++
+		if time.Since(t.visitedAt) > RecoveryCacheTTL {
+			ids = append(ids, id)
 		}
 	}
-
+	r.lock.RUnlock()
+	cleaned, err := r.tryExpire(ctx, ids...)
+	if err != nil {
+		r.lggr.Warnw("failed to clean visited upkeeps", "err", err)
+	}
 	if cleaned > 0 {
 		r.lggr.Debugw("gc: cleaned visited upkeeps", "cleaned", cleaned)
 	}
 }
+
+func (r *logRecoverer) tryExpire(ctx context.Context, ids ...string) (int, error) {
+	latestBlock, err := r.poller.LatestBlock(pg.WithParentCtx(ctx))
+	if err != nil {
+		return 0, fmt.Errorf("failed to get latest block: %w", err)
+	}
+	states, err := r.states.SelectByWorkIDsInRange(ctx, latestBlock-recoveryRescanBuffer, latestBlock, ids...)
```

Comment:
i am not fully sure if this range would be enough to look at? This should be atleast the range of blocks in every recovery interval? 

Maybe we should set a high value for recovery scan buffer?

Reply:
I think the similar comment also applies to the happy path recovery where we need to think about the window where we are looking for perform logs